src/controlplane/discoverer/discoverer.go

```diff
@@ -0,0 +1,93 @@
+package discoverer
+
+import (
+	"fmt"
+
+	"github.com/newrelic/infra-integrations-sdk/log"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+
+	"github.com/newrelic/nri-kubernetes/v2/internal/config"
+	"github.com/newrelic/nri-kubernetes/v2/internal/discovery"
+)
+
+// PodDiscoverer is used to discover control plane components.
+type PodDiscoverer interface {
+	// Discover returns looks for a pod mathching the selector, namespaces and
+	// node if matchNode is true.
+	Discover(config.AutodiscoverControlPlane) (*corev1.Pod, error)
+}
+
+type Config struct {
+	PodListerer discovery.PodListerer
+	NodeName    string
+}
+
+type OptionFunc func(c *ControlplanePodDiscoverer) error
+
+// WithLogger returns an OptionFunc to change the logger from the default noop logger.
+func WithLogger(logger log.Logger) OptionFunc {
+	return func(c *ControlplanePodDiscoverer) error {
+		c.logger = logger
+		return nil
+	}
+}
+
+// ControlplanePodDiscoverer implements PodDiscoverer interface.
+type ControlplanePodDiscoverer struct {
+	Config
+	logger log.Logger
+}
+
+// New returns an ControlplanePodDiscoverer.
+func New(config Config, opts ...OptionFunc) (*ControlplanePodDiscoverer, error) {
+	c := &ControlplanePodDiscoverer{
+		logger: log.Discard,
+		Config: config,
+	}
+
+	for i, opt := range opts {
+		if err := opt(c); err != nil {
+			return nil, fmt.Errorf("applying option #%d: %w", i, err)
+		}
+	}
+
+	return c, nil
+}
+
+// Discover returns the first Pod matching the namespace and selector from the listed pods.
+// If MatchNode is true the Pod must be running on the same node to match.
+//
+// Errors returned by this function should be managed as severe and not related to the
+// autodiscover entry. No error is returned if no Pod has been discovered.
+func (c *ControlplanePodDiscoverer) Discover(ad config.AutodiscoverControlPlane) (*corev1.Pod, error) {
+	podLister, ok := c.PodListerer.Lister(ad.Namespace)
+	if !ok {
+		return nil, fmt.Errorf("pod lister for namespace: %s not found", ad.Namespace)
+	}
+
+	labelsSet, err := labels.ConvertSelectorToLabelsMap(ad.Selector)
+	if err != nil {
+		return nil, fmt.Errorf("invalid selector %q: %w", ad.Selector, err)
+	}
+
+	selector := labels.SelectorFromSet(labelsSet)
+
+	pods, err := podLister.List(selector)
+	if err != nil {
+		return nil, fmt.Errorf("fail to list pods for selector: %v", labelsSet)
+	}
+
+	c.logger.Debugf("%d pods found with labels %q", len(pods), ad.Selector)
+
+	for _, pod := range pods {
+		if ad.MatchNode && pod.Spec.NodeName != c.Config.NodeName {
+			c.logger.Debugf("Discarding pod: %s running outside the node", pod.Name)
+			continue
+		}
+		// first pod matching all conditions is returned.
+		return pod, nil
+	}
+
+	return nil, nil
```

Comment:
Should we return a non-nil error here? Like a `ErrNoPodsFound`.

Reply:
yes it make sense! at least just to have this custom error 👍